File: crypto/pdpbatch/merkle.go
package pdpbatch

import (
	"bytes"
	"errors"
	"fmt"
	"sort"
)

type SingleProof struct {
	Index    int
	Siblings [][]byte
}

type NodeKey struct {
	Level    int
	Position int
}

type MultiProof struct {
	Depth      int
	ProofNodes map[NodeKey][]byte
}

func (p MultiProof) NodeCount() int {
	return len(p.ProofNodes)
}

type MerkleTree struct {
	ActualLeafCount int
	Size            int
	EmptyLeaf       []byte
	Levels          [][][]byte
	Root            []byte

	counter *OperationCounter
}

func NewMerkleTree(leaves [][]byte, counter *OperationCounter, emptyLeaf []byte) (*MerkleTree, error) {
	if len(leaves) == 0 {
		return nil, errors.New("merkle tree requires at least one leaf")
	}
	if emptyLeaf == nil {
		emptyLeaf = stableHashBytes(counter, "pad-leaf")
	}

	size := nextPowerOfTwo(len(leaves))
	padded := make([][]byte, 0, size)
	for _, leaf := range leaves {
		padded = append(padded, cloneBytes(leaf))
	}
	for len(padded) < size {
		padded = append(padded, cloneBytes(emptyLeaf))
	}

	levels := [][][]byte{padded}
	current := padded
	for len(current) > 1 {
		next := make([][]byte, 0, len(current)/2)
		for i := 0; i < len(current); i += 2 {
			next = append(next, stableHashBytes(counter, "merkle-node", current[i], current[i+1]))
		}
		levels = append(levels, next)
		current = next
	}

	return &MerkleTree{
		ActualLeafCount: len(leaves),
		Size:            size,
		EmptyLeaf:       cloneBytes(emptyLeaf),
		Levels:          levels,
		Root:            cloneBytes(levels[len(levels)-1][0]),
		counter:         counter,
	}, nil
}

func (t *MerkleTree) Depth() int {
	return len(t.Levels) - 1
}

func (t *MerkleTree) SingleProof(index int) (SingleProof, error) {
	if index < 0 || index >= t.ActualLeafCount {
		return SingleProof{}, fmt.Errorf("leaf index %d out of range", index)
	}

	siblings := make([][]byte, 0, t.Depth())
	position := index
	for level := 0; level < t.Depth(); level++ {
		siblingPosition := position ^ 1
		siblings = append(siblings, cloneBytes(t.Levels[level][siblingPosition]))
		position /= 2
	}
	return SingleProof{Index: index, Siblings: siblings}, nil
}

func (t *MerkleTree) MultiProof(indices []int) (MultiProof, error) {
	indexSet := make(map[int]struct{}, len(indices))
	for _, index := range indices {
		if index < 0 || index >= t.ActualLeafCount {
			return MultiProof{}, fmt.Errorf("leaf index %d out of range", index)
		}
		indexSet[index] = struct{}{}
	}
	if len(indexSet) == 0 {
		return MultiProof{}, errors.New("multi proof requires at least one index")
	}

	currentPositions := indexSet
	proofNodes := make(map[NodeKey][]byte)
	for level := 0; level < t.Depth(); level++ {
		positions := sortedKeys(currentPositions)
		nextPositions := make(map[int]struct{})
		for _, position := range positions {
			sibling := position ^ 1
			if _, ok := currentPositions[sibling]; !ok {
				proofNodes[NodeKey{Level: level, Position: sibling}] = cloneBytes(t.Levels[level][sibling])
			}
			nextPositions[position/2] = struct{}{}
		}
		currentPositions = nextPositions
	}

	return MultiProof{Depth: t.Depth(), ProofNodes: proofNodes}, nil
}

func VerifySingle(leafHash []byte, proof SingleProof, expectedRoot []byte, counter *OperationCounter) bool {
	current := cloneBytes(leafHash)
	position := proof.Index
	for _, sibling := range proof.Siblings {
		if position%2 == 0 {
			current = stableHashBytes(counter, "merkle-node", current, sibling)
		} else {
			current = stableHashBytes(counter, "merkle-node", sibling, current)
		}
		position /= 2
	}
	return bytes.Equal(current, expectedRoot)
}

func VerifyMulti(leafHashes map[int][]byte, proof MultiProof, expectedRoot []byte, counter *OperationCounter) bool {
	currentLevel := make(map[int][]byte, len(leafHashes))
	for index, hash := range leafHashes {
		currentLevel[index] = cloneBytes(hash)
	}

	for level := 0; level < proof.Depth; level++ {
		parentSet := make(map[int]struct{})
		for position := range currentLevel {
			parentSet[position/2] = struct{}{}
		}

		nextLevel := make(map[int][]byte, len(parentSet))
		for _, parentPosition := range sortedKeys(parentSet) {
			leftPosition := 2 * parentPosition
			rightPosition := leftPosition + 1

			leftHash, ok := currentLevel[leftPosition]
			if !ok {
				leftHash, ok = proof.ProofNodes[NodeKey{Level: level, Position: leftPosition}]
			}
			if !ok {
				return false
			}

			rightHash, ok := currentLevel[rightPosition]
			if !ok {
				rightHash, ok = proof.ProofNodes[NodeKey{Level: level, Position: rightPosition}]
			}
			if !ok {
				return false
			}

			nextLevel[parentPosition] = stableHashBytes(counter, "merkle-node", leftHash, rightHash)
		}
		currentLevel = nextLevel
	}

	root, ok := currentLevel[0]
	return ok && bytes.Equal(root, expectedRoot)
}

func sortedKeys(values map[int]struct{}) []int {
	keys := make([]int, 0, len(values))
	for key := range values {
		keys = append(keys, key)
	}
	sort.Ints(keys)
	return keys
}

func cloneBytes(value []byte) []byte {
	out := make([]byte, len(value))
	copy(out, value)
	return out
}

File: legacy/python_prototype/pdp_batch/merkle.py
"""
Merkle tree with single-proof and multi-proof authentication.

Paper symbol → Python entity:
  H1(i‖j‖σi,j)   → stable_hash_bytes(counter, "tag-leaf", i, j, σ)
  H1(Nl ‖ Nr)     → stable_hash_bytes(counter, "merkle-node", left, right)
  Li,j (leaf)     → bytes produced by _merkle_leaf() in protocol.py
  Γi   (root)     → MerkleTree.root
  Authi (SP mode) → SingleProof  – c independent sibling paths
  Authi (MP mode) → MultiProof   – compressed Tag-IMHT multi-proof
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, Iterable

from .algebra import OperationCounter, next_power_of_two, stable_hash_bytes


@dataclass
class SingleProof:
    index: int
    siblings: list[bytes]


@dataclass
class MultiProof:
    depth: int
    proof_nodes: dict[tuple[int, int], bytes]

    @property
    def node_count(self) -> int:
        return len(self.proof_nodes)


class MerkleTree:
    def __init__(self, leaves: list[bytes], counter: OperationCounter, empty_leaf: bytes | None = None) -> None:
        if not leaves:
            raise ValueError("Merkle tree requires at least one leaf")
        self.counter = counter
        self.actual_leaf_count = len(leaves)
        self.size = next_power_of_two(len(leaves))
        self.empty_leaf = empty_leaf or stable_hash_bytes(counter, "pad-leaf")
        padded = list(leaves) + [self.empty_leaf] * (self.size - len(leaves))
        self.levels = [padded]
        current = padded
        while len(current) > 1:
            nxt = []
            for index in range(0, len(current), 2):
                nxt.append(stable_hash_bytes(counter, "merkle-node", current[index], current[index + 1]))
            self.levels.append(nxt)
            current = nxt
        self.root = self.levels[-1][0]

    @property
    def depth(self) -> int:
        return len(self.levels) - 1

    def single_proof(self, index: int) -> SingleProof:
        if index >= self.actual_leaf_count:
            raise IndexError("leaf index out of range")
        siblings: list[bytes] = []
        position = index
        for level in range(self.depth):
            sibling_position = position ^ 1
            siblings.append(self.levels[level][sibling_position])
            position //= 2
        return SingleProof(index=index, siblings=siblings)

    def multi_proof(self, indices: Iterable[int]) -> MultiProof:
        index_set = sorted(set(indices))
        if not index_set:
            raise ValueError("multi proof requires at least one index")
        current_positions = set(index_set)
        proof_nodes: dict[tuple[int, int], bytes] = {}
        for level in range(self.depth):
            next_positions = set()
            for position in sorted(current_positions):
                sibling = position ^ 1
                if sibling not in current_positions:
                    proof_nodes[(level, sibling)] = self.levels[level][sibling]
                next_positions.add(position // 2)
            current_positions = next_positions
        return MultiProof(depth=self.depth, proof_nodes=proof_nodes)

    @staticmethod
    def verify_single(leaf_hash: bytes, proof: SingleProof, expected_root: bytes, counter: OperationCounter) -> bool:
        current = leaf_hash
        position = proof.index
        for sibling in proof.siblings:
            if position % 2 == 0:
                current = stable_hash_bytes(counter, "merkle-node", current, sibling)
            else:
                current = stable_hash_bytes(counter, "merkle-node", sibling, current)
            position //= 2
        return current == expected_root

    @staticmethod
    def verify_multi(
        leaf_hashes: Dict[int, bytes],
        proof: MultiProof,
        expected_root: bytes,
        counter: OperationCounter,
    ) -> bool:
        current_level = dict(leaf_hashes)
        for level in range(proof.depth):
            parent_positions = sorted({position // 2 for position in current_level})
            next_level: dict[int, bytes] = {}
            for parent_position in parent_positions:
                left_position = 2 * parent_position
                right_position = left_position + 1
                left_hash = current_level.get(left_position, proof.proof_nodes.get((level, left_position)))
                right_hash = current_level.get(right_position, proof.proof_nodes.get((level, right_position)))
                if left_hash is None or right_hash is None:
                    return False
                next_level[parent_position] = stable_hash_bytes(counter, "merkle-node", left_hash, right_hash)
            current_level = next_level
        return current_level.get(0) == expected_root
